### ISSUE TITLE
New blocks are rejected because of setHead operation

### ISSUE THREAD

**Issue by rjl493456442** [ethereum/go-ethereum#26693]
#### System information

Geth version: `geth version`
CL client & version: e.g. lighthouse/nimbus/prysm@v1.0.0
OS & Version: Windows/Linux/OSX
Commit hash : (if `develop`)

#### Expected behaviour

Whenever `SetHead` is performed, chain should be rewound to the specific position and import blocks on top smoothly.

#### Actual behaviour

`SetHead` may last for a long time, which locks the blockchain in `setHeadBeyondRoot` function. In the mean time consensus layer will keep feeding us new blocks via engine API. Specifically, in `newPayload` method, the new provided block will pass all checks(e.g. the parent block is existent, parent state is available, etc) and be blocked at `InsertBlockWithoutSetHead` which requires the blockchain lock.

When the `SetHead` is finished, the chain segment above the specified target is all removed, including the parent block of newly arrived payload in engine API. Eventually an `ErrUnknownAncestor = errors.New("unknown ancestor")` error will be returned which marks the new payload as invalid.

What's more, there is a mechanism in engine API to memory bad blocks to prevent handling them over and over again. Fortunately there is a time frame which gives the "bad block" another chance after some threshold. Currently the threshold is 128, it means after 128 attempts, the bad block will be gave another chance to import.

The issue is it's too much. Node needs to wait a long time to recover.

#### Steps to reproduce the behaviour

run `debug.SetHead()` when the node is already synced.

**Comment by MariusVanDerWijden** [comment]
The bigger issue might not be our own bad block cache, rather the bad block cache of the consensus layer. I've noticed that they might resend a block a handful of times (5) and stop sending it afterwards. So they seem to mark it as bad as well

**Comment by holiman** [comment]
> The bigger issue might not be our own bad block cache 

True, but that's not within scope of this specific ticket. The scenario described sounds like an error on our part. After a setHead, is  it correct to return an `ErrUnknownAncestor`? 

IMO, these two situation are semantically equivalent: 

- Node `A` syncs to `M`, then does setHead to back to `N` (say two weeks back). 
- Node `B` syncs to `N`, then is shut off. It is restarted two weeks later. 

So whatever one of them does, the other should do too.

**Comment by rjl493456442** [comment]
> After a setHead, is it correct to return an ErrUnknownAncestor?

Honestly, I think it's a correct behavior. We try to import a future block in this case.

**Comment by Francesreid** [comment]
Am I not going to get paid? Um yeh I've heard about all the scams people
do  Frances Reid
